### PR TITLE
Feature/karpenter

### PR DESCRIPTION
**Pull request by s0b01ev**

### CODE REVIEW

**Comment by s0b01ev** [comment]
I followed this migration tutorial steps:
https://karpenter.sh/docs/getting-started/migrating-from-cas/
First part, ending with [Add tags to subnets..](https://karpenter.sh/docs/getting-started/migrating-from-cas/#add-tags-to-subnets-and-security-groups) is implemented by cdk. Second part (starting with [Add tags to security groups.](https://karpenter.sh/docs/getting-started/migrating-from-cas/#add-tags-to-subnets-and-security-groups)) is in Makefile.
Karpeneter installation is optional now, and defined by KARPENTER env variable value ('true' or 'false')
`eksctl` and `envsubst` [(part of gettext-base package)](https://www.gnu.org/software/gettext/manual/html_node/envsubst-Invocation.html)  are prerequisites now

File: Makefile [hunk 1]
```diff
@@ -139,4 +147,53 @@ cert-gen: config-aws
 
 .PHONY: cert
 .SILENT: cert
-cert: cert-gen config
\ No newline at end of file
+cert: cert-gen config
+
+
```
Comment (arcanez) [inline Makefile:152]:
```suggestion
```

File: Makefile [hunk 2]
```diff
@@ -139,4 +147,53 @@ cert-gen: config-aws
...
+karpenter: karpenter-tag-sg karpenter-update-aws-auth karpenter-crd karpenter-helm karpenter-nodeclass
+
+.PHONY: karpenter-tag-sg
+karpenter-tag-sg:
+	@SECURITY_GROUPS=$(shell aws eks describe-cluster --name ${MDAI_CLUSTER_NAME} --query "cluster.resourcesVpcConfig.clusterSecurityGroupId" --profile ${AWS_PROFILE}) && \
+	aws ec2 create-tags --tags "Key=karpenter.sh/discovery,Value=${MDAI_CLUSTER_NAME}" --resources $${SECURITY_GROUPS} --profile ${AWS_PROFILE}
+
+.PHONY: karpenter-update-aws-auth
+karpenter-update-aws-auth:
+	eksctl create iamidentitymapping  --cluster DecisiveEngineCluster --username system:node:{{EC2PrivateDNSName}} \
```
Comment (arcanez) [inline Makefile:163]:
```suggestion
	eksctl create iamidentitymapping --cluster DecisiveEngineCluster --username system:node:{{EC2PrivateDNSName}} \
```

File: Makefile [hunk 3]
```diff
@@ -139,4 +147,53 @@ cert-gen: config-aws
...
+karpenter-helm:
+	@NODEGROUP=$(shell aws eks list-nodegroups --cluster-name "${MDAI_CLUSTER_NAME}" --query 'nodegroups[0]' --output text --profile ${AWS_PROFILE}) && \
+	helm upgrade -i karpenter oci://public.ecr.aws/karpenter/karpenter --version ${KARPENTER_VERSION} --namespace ${KARPENTER_NAMESPACE} \
+        --set settings.clusterName=${MDAI_CLUSTER_NAME} \
+        --set "serviceAccount.annotations.eks\.amazonaws\.com/role-arn=arn:aws:iam::${AWS_ACCOUNT}:role/KarpenterControllerRole-DecisiveEngineCluster" \
+        --set-json "affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution={\"nodeSelectorTerms\":[{\"matchExpressions\":[{\"key\":\"karpenter.sh/nodepool\",\"operator\":\"DoesNotExist\"}]},{\"matchExpressions\":[{\"key\":\"eks.amazonaws.com/nodegroup\",\"operator\":\"In\",\"values\":[\""$${NODEGROUP}"\"]}]}]}" \
+        --set controller.resources.requests.cpu=1 \
+        --set controller.resources.requests.memory=1Gi \
+        --set controller.resources.limits.cpu=1 \
+        --set controller.resources.limits.memory=1Gi
```
Comment (arcanez) [inline Makefile:184]: we may want to put this into a template and apply via `-f`

File: Makefile [hunk 4]
```diff
@@ -139,4 +147,53 @@ cert-gen: config-aws
...
+
+.PHONY: karpenter-nodeclass
+karpenter-nodeclass:
+	touch .k8s_version .arm_ami_parameter_name .arm_ami_id .amd_ami_parameter_name .amd_ami_id && \
+	aws eks describe-cluster --name ${MDAI_CLUSTER_NAME} --query cluster.version --out text --profile ${AWS_PROFILE} > .k8s_version && \
+	echo "/aws/service/eks/optimized-ami/`cat .k8s_version`/amazon-linux-2-arm64/recommended/image_id" > .arm_ami_parameter_name && \
+    echo "/aws/service/eks/optimized-ami/`cat .k8s_version`/amazon-linux-2/recommended/image_id" > .amd_ami_parameter_name && \
```
Comment (arcanez) [inline Makefile:191]:
```suggestion
	echo "/aws/service/eks/optimized-ami/`cat .k8s_version`/amazon-linux-2/recommended/image_id" > .amd_ami_parameter_name && \
```

File: Makefile [hunk 5]
```diff
@@ -139,4 +147,53 @@ cert-gen: config-aws
...
+	aws ssm get-parameter --name `cat .arm_ami_parameter_name` --query Parameter.Value --output text --profile ${AWS_PROFILE} > .arm_ami_id && \
+	aws ssm get-parameter --name `cat .amd_ami_parameter_name` --query Parameter.Value --output text --profile ${AWS_PROFILE} > .amd_ami_id
+	export ARM_AMI_ID=`cat .arm_ami_id`&&  export AMD_AMI_ID=`cat .amd_ami_id` && \
+	cat templates/karpenter-ec2nodeclass.yaml  | envsubst  > templates/karpenter-ec2nodeclass-values.yaml && \
```
Comment (arcanez) [inline Makefile:195]:
```suggestion
	cat templates/karpenter-ec2nodeclass.yaml | envsubst > templates/karpenter-ec2nodeclass-values.yaml && \
```

File: Makefile [hunk 6]
```diff
@@ -139,4 +147,53 @@ cert-gen: config-aws
...
+	rm .k8s_version .arm_ami_parameter_name .arm_ami_id .amd_ami_parameter_name .amd_ami_id
```
Comment (arcanez) [inline Makefile:196]:
```suggestion
	rm -f .k8s_version .arm_ami_parameter_name .arm_ami_id .amd_ami_parameter_name .amd_ami_id
```

**Review by arcanez — CHANGES_REQUESTED** [review]

File: Makefile (hunk 3 above)
Comment (s0b01ev) [reply Makefile:184]: well, as soon as 'envsubst' is required now, we can I guess

File: lib/decisive-engine-aws-cdk-stack.ts [hunk 7]
```diff
@@ -294,5 +301,212 @@ export class DecisiveEngineAwsCdkStack extends cdk.Stack {
     });
     mydecisiveEngineUi.node.addDependency(mdaiAppClient);
 
+    //
+    //    Karpenter
+    //
+    // steps (non-cdk way) taken from here
+    // https://karpenter.sh/docs/getting-started/migrating-from-cas/
+    if (config.KARPENTER.ENABLE === 'true') {
+
```
[the remark this answers is not in the document]
Comment (s0b01ev) [reply lib/decisive-engine-aws-cdk-stack.ts:310]: you suggest to remove conditional at all?

File: lib/decisive-engine-aws-cdk-stack.ts [hunk 8]
```diff
@@ -294,5 +301,212 @@ export class DecisiveEngineAwsCdkStack extends cdk.Stack {
...
+      const subnetsPriv = cluster.vpc.selectSubnets(
+          {
+            subnetType: ec2.SubnetType.PRIVATE
+          },
+      ).subnets;
+      const subnets = subnetsPub.concat(subnetsPriv)
+
+      cdk.Tags.of(this).add('karpenter.sh\/discovery', `${config.CLUSTER.NAME}`, {
+        includeResources: subnets,
```
[the remark this answers is not in the document]
Comment (s0b01ev) [reply lib/decisive-engine-aws-cdk-stack.ts:499]: We want to tag only subnets, that belong to the cluster.

File: lib/decisive-engine-aws-cdk-stack.ts [hunk 9]
```diff
@@ -294,5 +301,212 @@ export class DecisiveEngineAwsCdkStack extends cdk.Stack {
...
+          },
+      ).subnets;
+
+      const subnetsPriv = cluster.vpc.selectSubnets(
+          {
+            subnetType: ec2.SubnetType.PRIVATE
+          },
+      ).subnets;
+      const subnets = subnetsPub.concat(subnetsPriv)
+
```
[the remark this answers is not in the document]
Comment (s0b01ev) [reply lib/decisive-engine-aws-cdk-stack.ts:497]: Not sure what's the difference here. However with the suggested change, cluster.vpc  variable becomes detectable

File: lib/decisive-engine-aws-cdk-stack.ts [hunk 10]
```diff
@@ -294,5 +301,212 @@ export class DecisiveEngineAwsCdkStack extends cdk.Stack {
...
+      cdk.Tags.of(this).add('karpenter.sh\/discovery', `${config.CLUSTER.NAME}`, {
+        includeResources: subnets,
+      });
+
+      const vpcDefaultSecurityGroup = cluster.vpc.vpcDefaultSecurityGroup
+
```
[the remark this answers is not in the document]
Comment (s0b01ev) [reply lib/decisive-engine-aws-cdk-stack.ts:503]: Again, what the diff here?

File: lib/decisive-engine-aws-cdk-stack.ts [hunk 11]
```diff
@@ -1,85 +1,92 @@
...
       MDAI_COGNITO: {
-        UI_HOSTNAME: process.env.MDAI_UI_HOSTNAME || "console.mydecisive.ai",
-        USER_POOL_DOMAIN: process.env.MDAI_UI_USER_POOL_DOMAIN || "mydecisive",
-      }
+        UI_HOSTNAME: process.env.MDAI_UI_HOSTNAME || 'console.mydecisive.ai',
+        USER_POOL_DOMAIN: process.env.MDAI_UI_USER_POOL_DOMAIN || 'mydecisive',
+      },
+      KARPENTER: {
+        ENABLE: process.env.KARPENTER || 'true',
+        AWS_PARTITION: process.env.AWS_PARTITION || 'aws',
```
[the remark this answers is not in the document]
Comment (s0b01ev) [reply lib/decisive-engine-aws-cdk-stack.ts:73]: Do you think partition not needed at all? Can be only **aws** ?

**Pull request merged**